src/cogent3/app/typing.py

```diff
@@ -121,3 +121,28 @@ def hints_from_strings(*strings: Iterable[str]) -> list:
             raise ValueError(f"{string!r} not a known type constant")
         types.append(_mappings[string])
     return types
+
+
+def type_tree(hint, depth=0) -> tuple:
+    """compute the order of types"""
+    level_type = get_origin(hint)
+    if not level_type:
+        return depth + 1, hint
+
+    levels = []
+    depths = []
+    for arg in get_args(hint):
+        d, t = type_tree(arg, depth=depth)
+        levels.append(t)
+        depths.append(d)
+    depth = max(depths) + 1
```

Comment:
Can we use depths[-1] to find max depth? because I think the last is the max.

Reply:
For the following case `list[tuple[set[int]],set[int]]` the first element of `depths` is the maximum. So we have to use `max(depths)`